### PR TITLE
fix: pipeline name feature validation

### PR DESCRIPTION
**Pull request by Aravind-2023**
# Description

CI Pipeline Name didn't have proper validation and there was  toast popping up for invalid names instead for blank fields.

Fixes # [(AB279)](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/279/)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
All the testing have been done manually.
- [x] Tested the Pipeline name by providing invalid name.After providing invalid name I have whether the error is getting displayed below the Pipeline name field.
- [x] Tested whether the "some fields are missing" toast is only popped up when some fields are empty.Tested this by trying to click create pipeline with empty fields.

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 1]
```diff
@@ -317,7 +322,7 @@ export default function CIPipeline({
             )
         }
     }
-
+  console.log(formData.materials)
```
Comment (sohel-rp) [inline src/components/CIPipelineN/CIPipeline.tsx:325]: remove console log

File: src/components/ciPipeline/validationRules.ts [hunk 2]
```diff
@@ -1,16 +1,21 @@
...
-            return { message: 'This is required', isValid: false }
-        } else if (!re.test(value)) {
-            return { message: `Min of 3 characters; Start with lowercase; Use (a-z), (0-9), (-), (.)`, isValid: false }
-        } else {
-            return { message: null, isValid: true }
-        }
+        let re = PATTERNS.APP_NAME
+        let regExp = new RegExp(re)
+        let test = regExp.test(value)
+        if (value.length === 0) return { isValid: false, message: 'Please provide pipeline-name' }
```
Comment (sohel-rp) [inline src/components/ciPipeline/validationRules.ts:9]: why do we need so many variables here?

File: src/components/ciPipeline/validationRules.ts [hunk 3]
```diff
@@ -1,16 +1,21 @@
...
+        if (value.length < 2) return { isValid: false, message: 'Atleast 2 characters required' }
```
Comment (sohel-rp) [inline src/components/ciPipeline/validationRules.ts:10]: there should be space in `At least`

File: src/components/ciPipeline/validationRules.ts [hunk 4]
```diff
@@ -1,16 +1,21 @@
...
+        if (value.length > 50) return { isValid: false, message: 'Max 50 characters allowed' }
+        else if (!test)
```
Comment (sohel-rp) [inline src/components/ciPipeline/validationRules.ts:12]: use `regExp.test(value)` directly

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 5]
```diff
@@ -212,6 +213,10 @@ export default function CIPipeline({
         }
     }, [location.pathname])
 
+    useEffect(() => {
+        validateStage(activeStageName, formData)
+    }, [formData.name])
```
Comment (sohel-rp) [inline src/components/CIPipelineN/CIPipeline.tsx:218]: Remove this as we don't need useEffect for validating name. Put it in onChange function itself.

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 6]
```diff
@@ -378,10 +382,22 @@ export default function CIPipeline({
             !formDataErrorObj.preBuildStage.isValid ||
             !formDataErrorObj.postBuildStage.isValid
         ) {
+            let branchNamePresent = true
+
             setLoadingData(false)
-            toast.error('Some Required Fields are missing')
+
+            branchNamePresent = formData.materials?.reduce((branchNamePresent, obj) => {
```
Comment (sohel-rp) [inline src/components/CIPipelineN/CIPipeline.tsx:389]: Please update as discussed

File: src/components/ciPipeline/validationRules.ts [hunk 7]
```diff
@@ -1,16 +1,19 @@
...
     name = (value: string): { message: string | null; isValid: boolean } => {
-        const re = new RegExp('^[a-z][a-z0-9-.]+[a-z0-9]$')
-        if (value && value.length < 1) {
-            return { message: 'This is required', isValid: false }
-        } else if (!re.test(value)) {
-            return { message: `Min of 3 characters; Start with lowercase; Use (a-z), (0-9), (-), (.)`, isValid: false }
-        } else {
-            return { message: null, isValid: true }
-        }
+        let regExp = new RegExp(PATTERNS.APP_NAME)
```
Comment (sohel-rp) [inline src/components/ciPipeline/validationRules.ts:6]: Make it const

Comment (sohel-rp) [reply src/components/ciPipeline/validationRules.ts:6]: Also, format the changes

File: src/components/cdPipeline/validationRules.ts [hunk 8]
```diff
@@ -2,20 +2,29 @@ import { PATTERNS } from "../../config";
...
-            };
+    name = (value: string, pattern?: string): { isValid: boolean, message: string } => {
+        let re="";
+        if(pattern){
+        re=pattern
         }
-        else return { isValid: false, message: 'This is a required field' };
+        else{
+         re = PATTERNS.APP_NAME
+        }
```
Comment (sohel-rp) [inline src/components/cdPipeline/validationRules.ts:12]: We don't need this as we were already doing it with `new RegExp(pattern || PATTERNS.CD_PIPELINE_NAME)`. Please revert this and also confirm if we want all app name validations for cd pipeline name.

File: src/components/cdPipeline/CDPipeline.tsx [hunk 9]
```diff
@@ -575,11 +599,10 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
             this.validationRules.name(this.state.pipelineConfig.name).isValid &&
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:599]: Move this validation before any computation

Comment (sohel-rp) [reply src/components/cdPipeline/CDPipeline.tsx:599]: As we're already checking for pipeline name & namespace above it makes sense to merge both

File: src/components/cdPipeline/CDPipeline.tsx [hunk 10]
```diff
@@ -494,13 +505,15 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
     }
 
     handlePipelineName = (event) => {
-        let { pipelineConfig } = { ...this.state }
+        const { pipelineConfig, errorForm } = { ...this.state }
+        const _errorForm = { ...errorForm }
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:509]: we don't need this extra const use extracted errorForm directly

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 11]
```diff
@@ -379,9 +378,15 @@ export default function CIPipeline({
             !formDataErrorObj.postBuildStage.isValid
         ) {
             setLoadingData(false)
-            toast.error('Some Required Fields are missing')
+            const checkBranch = (obj) => obj.value === ''
+            let branchNameNotPresent = formData.materials.some(checkBranch)
+            if (formData.name === '' || branchNameNotPresent) {
```
Comment (sohel-rp) [inline src/components/CIPipelineN/CIPipeline.tsx:383]: Are we not able to use existing states/values?

File: src/components/CIPipelineN/Build.tsx [hunk 12]
```diff
@@ -188,6 +190,9 @@ export function Build({
         const _form = { ...formData }
         _form.name = event.target.value
         setFormData(_form)
+         const _formDataErrorObj={...formDataErrorObj}
+         _formDataErrorObj.name=validationRules.name(_form.name)
+         setFormDataErrorObj(_formDataErrorObj);
```
Comment (sohel-rp) [inline src/components/CIPipelineN/Build.tsx:195]: fix indentation

File: src/components/CIPipelineN/CIPipeline.tsx (hunk 11 above)
Comment (sohel-rp) [reply src/components/CIPipelineN/CIPipeline.tsx:383]: Also, you can simplify checkBranch something like,
```
formData.materials.some((_mat) => !_mat.value)
```

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 13]
```diff
@@ -38,6 +38,7 @@ import { ReactComponent as AlertTriangle } from '../../assets/icons/ic-alert-tri
 import { getModuleInfo } from '../v2/devtronStackManager/DevtronStackManager.service'
 import { ModuleStatus } from '../v2/devtronStackManager/DevtronStackManager.type'
 
+
```
Comment (shivani170) [inline src/components/CIPipelineN/CIPipeline.tsx:41]: Please remove this extra line

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 14]
```diff
@@ -379,9 +378,14 @@ export default function CIPipeline({
             !formDataErrorObj.postBuildStage.isValid
         ) {
             setLoadingData(false)
-            toast.error('Some Required Fields are missing')
+            let branchNameNotPresent = formData.materials.some((_mat) => !_mat.value)
+            if (formData.name === '' || branchNameNotPresent) {
+                toast.error('Some required fields are missing')
```
Comment (shivani170) [inline src/components/CIPipelineN/CIPipeline.tsx:383]: Please pick that message from  constant messaging `MULTI_REQUIRED_FIELDS_MSG`

File: src/components/ciPipeline/validationRules.ts [hunk 15]
```diff
@@ -1,16 +1,19 @@
...
+        if (value.length === 0) return { isValid: false, message: 'This is required' }
+        if (value.length < 2) return { isValid: false, message: 'At least 2 characters required' }
+        if (value.length > 50) return { isValid: false, message: 'Max 50 characters allowed' }
+        else if (!regExp.test(value))
+            return {
+                isValid: false,
+                message:
+                    "Min 2 chars; Start with alphabet; End with alphanumeric; Use only lowercase; Allowed:(-), (.); Do not use 'spaces'",
```
Comment (shivani170) [inline src/components/ciPipeline/validationRules.ts:14]: Extract out from constant messaging

File: src/components/cdPipeline/validationRules.ts [hunk 16]
```diff
@@ -1,30 +1,31 @@
...
+        if (value.length > 50) return { isValid: false, message: 'Max 50 characters allowed' }
+        else if (!regExp.test(value))
+            return {
                 isValid: false,
-                message: `Must follow the pattern ${PATTERNS.CD_PIPELINE_NAME}`
-            };
-        }
-        else return { isValid: false, message: 'This is a required field' };
+                message:
+                    "Min 2 chars; Start with alphabet; End with alphanumeric; Use only lowercase; Allowed:(-), (.); Do not use 'spaces'",
```
Comment (shivani170) [inline src/components/cdPipeline/validationRules.ts:13]: Extract out from constant messaging

File: src/components/cdPipeline/CDPipeline.tsx [hunk 17]
```diff
@@ -1392,9 +1416,9 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
                                     : `auto`,
 
                             maxHeight:
-                            this.props.match.params.cdPipelineId || this.state.isAdvanced
-                                ? `auto`
-                                : `calc(100vh - 164px)`,
+                                this.props.match.params.cdPipelineId || this.state.isAdvanced
+                                    ? `auto`
```
Comment (shivani170) [inline src/components/cdPipeline/CDPipeline.tsx:1420]: Remove interpolation use double quotes instaed

File: src/components/cdPipeline/CDPipeline.tsx [hunk 18]
```diff
@@ -728,11 +749,12 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
 
     renderHeader() {
-        const title = this.isWebhookCD && this.props.match.params.workflowId === '0'
-            ? 'Deploy image from external source'
-            : this.props.match.params.cdPipelineId
-            ? 'Edit deployment pipeline'
-            : 'Create deployment pipeline'
+        const title =
+            this.isWebhookCD && this.props.match.params.workflowId === '0'
+                ? 'Deploy image from external source'
```
Comment (shivani170) [inline src/components/cdPipeline/CDPipeline.tsx:754]: Extract out from constant messaging

File: src/components/cdPipeline/CDPipeline.tsx [hunk 19]
```diff
@@ -549,7 +561,28 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
 
     savePipeline() {
-        this.setState({ showError: true, loadingData: true })
+        const { pipelineConfig, errorForm } = { ...this.state }
+        errorForm.pipelineNameError = this.validationRules.name(pipelineConfig.name)
+        errorForm.nameSpaceError = this.validationRules.namespace(pipelineConfig.namespace)
+        errorForm.envNameError = this.validationRules.environment(pipelineConfig.environmentId)
+        this.setState({ errorForm })
+        let valid =
+            !!!this.state.pipelineConfig.environmentId &&
```
Comment (shivani170) [inline src/components/cdPipeline/CDPipeline.tsx:570]: Make it  !!this.state.pipelineConfig.environmentId

File: src/components/CIPipelineN/CIPipeline.tsx [hunk 20]
```diff
@@ -37,6 +37,7 @@ import { Build } from './Build'
 import { ReactComponent as AlertTriangle } from '../../assets/icons/ic-alert-triangle.svg'
 import { getModuleInfo } from '../v2/devtronStackManager/DevtronStackManager.service'
 import { ModuleStatus } from '../v2/devtronStackManager/DevtronStackManager.type'
+import { MULTI_REQUIRED_FIELDS_MSG } from '../../config/constantMessaging'
```
Comment (shivani170) [inline src/components/CIPipelineN/CIPipeline.tsx:40]: Aren't we using it anywhere if not then please remove

File: src/components/cdPipeline/CDPipeline.tsx [hunk 21]
```diff
@@ -435,11 +444,12 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
             pipelineConfig.isClusterCdActive = selection.isClusterCdActive
             pipelineConfig.runPreStageInEnv = pipelineConfig.isClusterCdActive && pipelineConfig.runPreStageInEnv
             pipelineConfig.runPostStageInEnv = pipelineConfig.isClusterCdActive && pipelineConfig.runPostStageInEnv
-            this.setState({ environments: list, pipelineConfig }, () => {
+            this.setState({ environments: list, pipelineConfig, errorForm }, () => {
                 getConfigMapAndSecrets(this.props.match.params.appId, this.state.pipelineConfig.environmentId)
                     .then((response) => {
                         this.configMapAndSecrets = response.result
-                        this.setState({ view: ViewType.FORM })
+
```
Comment (shivani170) [inline src/components/cdPipeline/CDPipeline.tsx:451]: remove this line

File: src/components/cdPipeline/CDPipeline.tsx [hunk 22]
```diff
@@ -728,11 +749,12 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
-        const title = this.isWebhookCD && this.props.match.params.workflowId === '0'
-            ? 'Deploy image from external source'
-            : this.props.match.params.cdPipelineId
-            ? 'Edit deployment pipeline'
-            : 'Create deployment pipeline'
+        const title =
+            this.isWebhookCD && this.props.match.params.workflowId === '0'
+                ? 'Deploy image from external source'
+                : this.props.match.params.cdPipelineId
+                ? 'Edit deployment pipeline'
```
Comment (shivani170) [inline src/components/cdPipeline/CDPipeline.tsx:756]: Extract from constant

**Review by shivani170 — APPROVED** [review]

**Review by milindk8 — APPROVED** [review]

File: src/components/cdPipeline/CDPipeline.tsx [hunk 23]
```diff
@@ -549,7 +566,27 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
     savePipeline() {
-        this.setState({ showError: true, loadingData: true })
+        const { pipelineConfig, errorForm } = { ...this.state }
+        errorForm.pipelineNameError = this.validationRules.name(pipelineConfig.name)
+        errorForm.nameSpaceError = this.validationRules.namespace(pipelineConfig.namespace)
+        errorForm.envNameError = this.validationRules.environment(pipelineConfig.environmentId)
+        this.setState({ errorForm })
+        let valid =
+            !!this.state.pipelineConfig.environmentId &&
+            this.validationRules.name(this.state.pipelineConfig.name).isValid &&
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:576]: Can't we use pipelineNameError from 570?

Comment (sohel-rp) [reply src/components/cdPipeline/CDPipeline.tsx:576]: Also, we've already destructured `pipelineConfig` from state so can't use that?

**Review by sohel-rp — APPROVED** [review]

**Pull request merged**